### PR TITLE
test pre execution r/w conflicts 

### PR DESCRIPTION
**Pull request by yontyon**
* Adding functionality to the tracker to produce conflicted bft requests.
* Adding test to pre execution tests for r/w conflicts

### CODE REVIEW

File: tests/apollo/test_skvbc_preexecution.py [hunk 1]
```diff
@@ -366,6 +366,30 @@ async def test_alternate_f_isolated(self, bft_network, tracker):
             last_block = await tracker.get_last_block_id(read_client)
             assert last_block > start_block
 
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    @verify_linearizability(pre_exec_enabled=True)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_preexecution.py:371]: let's explicitlhy state that `no_conflicts=False` here

File: tests/apollo/test_skvbc_preexecution.py [hunk 2]
```diff
@@ -366,6 +366,30 @@ async def test_alternate_f_isolated(self, bft_network, tracker):
...
+        try:
+            with trio.move_on_after(seconds=30):
+                await tracker.run_concurrent_conflict_ops(ops, concurrency_level=2, write_weight=1)
+        except trio.TooSlowError:
+            pass
+
+        last_block = await tracker.get_last_block_id(read_client)
+
+        # We produced at least one conflict.
+        assert last_block < start_block + ops
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_preexecution.py:392]: As an experiment, can you try running the same test by setting `no_conflicts=True`, and making sure this assert fails?

File: tests/apollo/test_skvbc_preexecution.py [hunk 3]
```diff
@@ -366,6 +366,30 @@ async def test_alternate_f_isolated(self, bft_network, tracker):
...
+        bft_network.start_all_replicas()
+
+        read_client = bft_network.random_client()
+        start_block = await tracker.get_last_block_id(read_client)
+
+        ops = 20
+
+        try:
+            with trio.move_on_after(seconds=30):
+                await tracker.run_concurrent_conflict_ops(ops, concurrency_level=2, write_weight=1)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_preexecution.py:385]: why concurrency_level is 2 ?

File: tests/apollo/test_skvbc_preexecution.py [hunk 4]
```diff
@@ -366,6 +366,30 @@ async def test_alternate_f_isolated(self, bft_network, tracker):
...
             assert last_block > start_block
 
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    @verify_linearizability(pre_exec_enabled=True)
+    async def test_conflicting_requests(self, bft_network, tracker):
+        """
+        Launch pre-process conflicting request and make sure that conflicting requests are not committed
+        """
+        bft_network.start_all_replicas()
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_preexecution.py:376]: once this test is stable, let's add another one where the conflict happens in the presence of _f_ crashed replicas

File: tests/apollo/test_skvbc_preexecution.py (hunk 1 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_preexecution.py]: done

File: tests/apollo/test_skvbc_preexecution.py (hunk 2 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_preexecution.py:392]: done

File: tests/apollo/test_skvbc_preexecution.py (hunk 3 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_preexecution.py]: fixed

File: tests/apollo/util/skvbc_history_tracker.py [hunk 5]
```diff
@@ -912,6 +915,32 @@ async def run_concurrent_ops(self, num_ops, write_weight=.70):
             sent += len(clients)
         return read_count, write_count
 
+    async def run_concurrent_conflict_ops(self, num_ops, write_weight=.70):
```
Comment (teoparvanov) [inline tests/apollo/util/skvbc_history_tracker.py:918]: Why do you need this additional method? Isn't the `no_conflicts` flag sufficient to generate conflicts?

Comment (yontyon) [reply tests/apollo/util/skvbc_history_tracker.py:918]: no_concflict=False **may** generate conflicts.
In this test we want to generate conflict for sure

Comment (teoparvanov) [reply tests/apollo/util/skvbc_history_tracker.py:918]: hmm, you are right, but can you at least try to reduce the code duplication between `run_concurrent_ops()` and `run_concurrent_conflict_ops()` ?

File: tests/apollo/test_skvbc_preexecution.py (hunk 4 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_preexecution.py:374]: done

File: tests/apollo/util/skvbc_history_tracker.py (hunk 5 above)
Comment (yontyon) [reply tests/apollo/util/skvbc_history_tracker.py:904]: done

Comment (yontyon) [reply tests/apollo/util/skvbc_history_tracker.py:904]: done

File: tests/apollo/util/skvbc_history_tracker.py [hunk 6]
```diff
@@ -1074,6 +1099,32 @@ async def run_concurrent_ops(self, num_ops, write_weight=.70):
...
+            async with trio.open_nursery() as nursery:
+                for client in clients:
+                    if random.random() < write_weight:
+                        nursery.start_soon(self.send_tracked_kv_set, client, readset, writeset, read_version)
+                        write_count += 1
+                    else:
+                        nursery.start_soon(self.send_tracked_read, client, max_size)
+                        read_count += 1
+            sent += len(clients)
+        return read_count, write_count
```
Comment (teoparvanov) [inline tests/apollo/util/skvbc_history_tracker.py:1126]: why not do the same refactoring here too?

Comment (teoparvanov) [reply tests/apollo/util/skvbc_history_tracker.py:1126]: I mean, to extract `send_concurrent_ops()`...

File: tests/apollo/test_skvbc_preexecution.py (hunk 4 above)
Comment (teoparvanov) [reply tests/apollo/test_skvbc_preexecution.py:374]: thanks, almost there - see my comment below...

File: tests/apollo/util/skvbc_history_tracker.py (hunk 5 above)
Comment (teoparvanov) [reply tests/apollo/util/skvbc_history_tracker.py:904]: thanks, almost there - see my comment below...

File: tests/apollo/util/skvbc_history_tracker.py (hunk 6 above)
Comment (yontyon) [reply tests/apollo/util/skvbc_history_tracker.py]: done

**Pull request merged**